File: src/com/cattailsw/nanidroid/ShioriFactory.java
package com.cattailsw.nanidroid;

import com.cattailsw.nanidroid.shiori.EchoShiori;
import com.cattailsw.nanidroid.shiori.SatoriPosixShiori;
import com.cattailsw.nanidroid.shiori.Shiori;
import com.cattailsw.nanidroid.shiori.NanidroidShiori;
import com.cattailsw.nanidroid.shiori.NotSupportedShiori;
import com.cattailsw.nanidroid.shiori.Kawari;
import java.io.File;
import java.util.Map;
import android.content.Context;

public class ShioriFactory {

    private static ShioriFactory _self = null;
    private ShioriFactory() {

    }

    public static final ShioriFactory getInstance() {
	if ( _self == null )
	    _self = new ShioriFactory();

	return _self;
    }

    // path is the master ghost path 



    public Shiori getShiori(String path, Map<String, String> masterDesc){
	return getShiori(path, masterDesc, null);
    }

    private Shiori checkShioriByPath(String path, Context ctx) {
	// need to do some more checking... argh
	if ( (new File(path, "kawarirc.kis")).exists() ) {
	    return new Kawari(path); // should be kawari878
	}
	else if ( (new File(path, "kawari.ini") ).exists() ) {
	    return new NotSupportedShiori(ctx); // kawari 7 or something like this
	}
	else if ( (new File(path, "aya5.txt")).exists()) {
	    return new NotSupportedShiori(ctx); // assume yaya
	}

	return new NotSupportedShiori(ctx);
    }

    public Shiori getShiori(String path, Map<String, String> masterDesc, Context ctx) {
	// need to first read the shiori dll file name from descript.txt
	String sdllname = masterDesc.get("shiori");
	if ( sdllname == null ){
	    // check path for files...
	    return checkShioriByPath(path, ctx);
	}
	else {
	    if ( sdllname.matches("Nanidroid") ) {
		return new NanidroidShiori(ctx, path);
	    }
	    else if ( sdllname.matches("satori.dll") ) {
		return new SatoriPosixShiori(path);
	    }
	    else if ( sdllname.matches("shiori.dll") ) {
		return checkShioriByPath(path, ctx);
	    }
	    else if ( sdllname.matches("yaya.dll") ) {
		return new NotSupportedShiori(ctx); // should be yaya
	    }
	}

	return new NotSupportedShiori(ctx);
    }

}

File: app/src/main/java/com/cattailsw/nanidroid/ShioriFactory.kt
package com.cattailsw.nanidroid

import android.content.Context
import com.cattailsw.nanidroid.shiori.*
import java.io.File

class ShioriFactory private constructor() {
    companion object {
        private var _self: ShioriFactory? = null

        @JvmStatic
        fun getInstance(): ShioriFactory {
            if (_self == null) {
                _self = ShioriFactory()
            }
            return _self!!
        }
    }

    fun getShiori(path: String, masterDesc: Map<String, String>): Shiori {
        return getShiori(path, masterDesc, null)
    }

    private fun checkShioriByPath(path: String, ctx: Context?): Shiori {
        return if (File(path, "kawarirc.kis").exists()) {
            Kawari(path)
        } else if (File(path, "kawari.ini").exists()) {
            NotSupportedShiori(ctx)
        } else if (File(path, "aya5.txt").exists()) {
            NotSupportedShiori(ctx)
        } else {
            NotSupportedShiori(ctx)
        }
    }

    fun getShiori(path: String, masterDesc: Map<String, String>, ctx: Context?): Shiori {
        val sdllname = masterDesc["shiori"]
        return if (sdllname == null) {
            checkShioriByPath(path, ctx)
        } else {
            if (sdllname.matches("Nanidroid".toRegex())) {
                NanidroidShiori(ctx!!, path)
            } else if (sdllname.matches("satori.dll".toRegex())) {
                SatoriPosixShiori(path)
            } else if (sdllname.matches("shiori.dll".toRegex())) {
                checkShioriByPath(path, ctx)
            } else if (sdllname.matches("yaya.dll".toRegex())) {
                NotSupportedShiori(ctx)
            } else {
                NotSupportedShiori(ctx)
            }
        }
    }
}
